proposals/explicit-backing-fields.md

```diff
@@ -0,0 +1,424 @@
+```kotlin
+val counter: Int
+    field = atomic(0)
+    get() = field.value
+```
+
+When explicit backing field with type `F` for a property with type `P` is declared,
+then compiler can derive getter and, for `var` properties, setter implementation automatically:
+
+* If `F` is a subtype of `P`, the compiler can provide a default getter.
+* If `P` is a subtype of `F`, the compiler can provide a default setter.
+
+If the compiler can not provide a getter, the user must declare it explicitly.
+The same applies to setters in case of `var` properties.
+
+```kotlin
+val city: String
+    field = MutableLiveData<String>()
+    get() = field.value ?: "-" // It is an error if getter is not explicitly specified here
+```
+
+### Explicit backing field in combination with property initializer
+
+It's allowed to have both property initializer and explicit backing field specified.
+
+```kotlin
+val city: StateFlow<String> = field.asStateFlow()
+    field = MutableStateFlow("")
+```
+
+In this case, the property initializer expression (`field.asStateFlow()` in example above) is evaluated once
+immediately after the backing field is initialized and then returned every time the getter of property is called.
+
+If property with explicit backing field has initializer, it's prohibited to declare accessors.
+
+### Accessing field inside accessors and initializers
+
+It's possible to access explicit backing field by calling `field` from inside accessors (like it works now for ordinary
+properties) and property initialization expression.
+
+Backing field assignability is the same as it is now for `field` references in getters and setters:
+
+* `var` properties have mutable backing fields.
+* `val` properties have read-only backing fields. That is, assignment to `field` inside a `val` property getter results in an error.
+
+The following additional restrictions apply:
+
+* It's prohibited to reference `field` inside explicit backing field initializer expression.
+* Explicit backing field must be referenced through `field` in property initializer, setter and getter (to prevent feature abuse).
+
+### Type inference
+
+Explicit type declaration on explicit backing field is optional. If it is not specified, it's inferred from the explicit backing field initialization expression.
+
+Explicit type declaration on property, which has explicit backing field, is optional. If it is not specified, it's inferred from
+the property initialization or from getter. If none of them are specified, the property type is equal to the explicit backing field type.
+
+It's a warning if the property type is equal to the explicit backing field type.
+
+### Other restrictions
+
+There are the following additional semantic restrictions on the property declaration.
+Property with explicit backing field:
+
+* must be final (otherwise calling a field instead of a getter would be undesirable behavior)
+* can't be `const`
+* can't be `expect` or `external`
+* can't be `inline` (but its accessors can)
```

Comment:
FTR: an `inline` property means its accessors are `inline`.

Why do we want to have `inline` accessors for EBF properties? This means that the EBF leaks into the public ABI, has to be accessible from the outside and I'm not sure I like this.

Atm, `inline` properties cannot have backing fields, and I do not understand why we want to relax the restriction here.

Reply:
Smth I originally forgot -- do we say anywhere that EBF are prohibited for extension properties? =)